In [26]:
import nltk

#nltk.download("wordnet")
#nltk.download("wordnet_ic")
    
from nltk.corpus import wordnet

[nltk_data] Downloading package wordnet_ic to
[nltk_data]     C:\Users\SoDipole\AppData\Roaming\nltk_data...
[nltk_data]   Unzipping corpora\wordnet_ic.zip.


<b>1) Найти все значения (синсеты) для лексемы right<b>

In [9]:
right = wordnet.synsets('right')
for ss in right:
    print(ss, ss.definition())

Synset('right.n.01') an abstract idea of that which is due to a person or governmental body by law or tradition or nature; it is something that nobody can take away"
Synset('right.n.02') location near or direction toward the right side; i.e. the side to the south when a person or object faces east
Synset('right_field.n.01') the piece of ground in the outfield on the catcher's right
Synset('right.n.04') those who support political or social or economic conservatism; those who believe that things are better left unchanged
Synset('right.n.05') the hand that is on the right side of the body
Synset('right.n.06') a turn toward the side of the body that is on the south when the person is facing east
Synset('right.n.07') anything in accord with principles of justice
Synset('right.n.08') (frequently plural) the interest possessed by law or custom in some intangible thing
Synset('right.v.01') make reparations or amends for
Synset('right.v.02') put in or restore to an upright position
Synset('rig

<b>2) Найти определение для лексемы right в значении (а) "право (человека)" и в значении (b) "право (направление)"</b>

In [11]:
print(right[0], right[0].definition())
print(right[1], right[1].definition())

Synset('right.n.01') an abstract idea of that which is due to a person or governmental body by law or tradition or nature; it is something that nobody can take away"
Synset('right.n.02') location near or direction toward the right side; i.e. the side to the south when a person or object faces east


<b>3) Найдите два произвольных контекста для слова right в значениях (a) "право (человека)" и (b) "право (направление)"; продемонстрируйте на них действие алгоритма Леска для разрешения неоднозначности</b>

In [14]:
sent1 = 'You have a right to say what you please.'
sent2 = 'You need to go straight down the hall and then turn to the right.'
sent1_tokens = [word.strip('.,') for word in sent1.split(' ')]
sent2_tokens = [word.strip('.,') for word in sent2.split(' ')]
print(sent1_tokens)
print(sent2_tokens)

['You', 'have', 'a', 'right', 'to', 'say', 'what', 'you', 'please']
['You', 'need', 'to', 'go', 'straight', 'down', 'the', 'hall', 'and', 'then', 'turn', 'to', 'the', 'right']


In [15]:
from nltk.wsd import lesk
print(lesk(sent1_tokens, 'right').definition())
print(lesk(sent2_tokens, 'right').definition())

location near or direction toward the right side; i.e. the side to the south when a person or object faces east
location near or direction toward the right side; i.e. the side to the south when a person or object faces east


Алгоритм Леска ошибся в первом случае. Попробуем ограничить часть речи существительными.

In [16]:
print(lesk(sent1_tokens, 'right', 'n').definition())
print(lesk(sent2_tokens, 'right', 'n').definition())

location near or direction toward the right side; i.e. the side to the south when a person or object faces east
location near or direction toward the right side; i.e. the side to the south when a person or object faces east


Не помогло :(

<b>4) Найдите гиперонимы для значения (a) и гиперонимы для значения (b)</b>

In [49]:
for ss in right[0].hypernyms():
    print(ss, ss.definition())
for ss in right[1].hypernyms():
    print(ss, ss.definition())

Synset('abstraction.n.01') a concept or idea not associated with any specific instance
Synset('position.n.01') the particular portion of space occupied by something


<b>5) Вычислите наименьшее расстояние между значением right "право (человека)" и значениями лексемы industry, а также right "право (направление)" и значениями лексемы leaf.
Найти min (d(plant: "право (человека)", industry), d(plant: "право (человека)", leaf)), а также min (d(plant: "право (направление)", industry), d(plant: "право (направление)", leaf))</b>

In [20]:
freedom = wordnet.synsets('freedom')
direction = wordnet.synsets('direction')

def get_dist_sim(ss1, lexeme):
    distances = []
    similarities = []
    for ss2 in lexeme:
        dist = ss1.shortest_path_distance(ss2)
        if dist is not None:
            distances.append(dist)
            sim = ss1.path_similarity(ss2)
            similarities.append(sim)
    return distances, similarities

# min d(right: "право (человека)", freedom)
dist1 = get_dist_sim(right[0], freedom)[0]
print('min d(right: "право (человека)", freedom): {}'.format(min(dist1)))
print('closest lemma definition: {}\n'.format(freedom[dist1.index(min(dist1))].definition()))

# min d(right: "право (человека)", direction)
dist2 = get_dist_sim(right[0], direction)[0]
print('min d(right: "право (человека)", direction): {}'.format(min(dist2)))
print('closest lemma definition: {}\n'.format(direction[dist2.index(min(dist2))].definition()))

# min d(right: "право (направление)", freedom)
dist3 = get_dist_sim(right[1], freedom)[0]
print('min d(right: "право (направление)", freedom): {}'.format(min(dist3)))
print('closest lemma definition: {}\n'.format(freedom[dist3.index(min(dist3))].definition()))

# min d(right: "право (направление)", direction)
dist4 = get_dist_sim(right[1], direction)[0]
print('min d(right: "право (направление)", direction): {}'.format(min(dist4)))
print('closest lemma definition: {}\n'.format(direction[dist4.index(min(dist4))].definition()))

print('min (d(right: "право (человека)", freedom), d(right: "право (человека)", direction)): {}'.format(min(min(dist1), min(dist2))))
print('min (d(right: "право (направление)", freedom), d(right: "право (направление)", direction)): {}'.format(min(min(dist3), min(dist4))))

min d(right: "право (человека)", freedom): 10
closest lemma definition: the condition of being free; the power to act or speak or think without externally imposed restraints

min d(right: "право (человека)", direction): 8
closest lemma definition: a general course along which something has a tendency to develop

min d(right: "право (направление)", freedom): 10
closest lemma definition: the condition of being free; the power to act or speak or think without externally imposed restraints

min d(right: "право (направление)", direction): 6
closest lemma definition: a line leading to a place or point

min (d(right: "право (человека)", freedom), d(right: "право (человека)", direction)): 8
min (d(right: "право (направление)", freedom), d(right: "право (направление)", direction)): 6


Кратчайшее расстояние от права (человека) до freedom - 10, до direction - 8. Получается, что до direction расстояние короче, что кажется странным. Но если посмотреть на определение direction в этом случае, то вполне логично.

Кратчайшее расстояние от права (направления) до freedom - 10, до direction - 6. До direction расстояние короче, слава богу.

Если имелось в виду посчитать не расстояние, а сходство путей на основе кратчайшего расстояния (минимальный путь --> максимальное сходство), вот оно:

In [22]:
# max s(right: "право (человека)", freedom)
sim1 = get_dist_sim(right[0], freedom)[1]
print('max s(right: "право (человека)", freedom): {}'.format(max(sim1)))
print('closest lemma definition: {}\n'.format(freedom[sim1.index(max(sim1))].definition()))

# max s(right: "право (человека)", direction)
sim2 = get_dist_sim(right[0], direction)[1]
print('max s(right: "право (человека)", direction): {}'.format(max(sim2)))
print('closest lemma definition: {}\n'.format(direction[sim2.index(max(sim2))].definition()))

# max s(right: "право (направление)", freedom)
sim3 = get_dist_sim(right[1], freedom)[1]
print('max s(right: "право (направление)", freedom): {}'.format(min(sim3)))
print('closest lemma definition: {}\n'.format(freedom[sim3.index(max(sim3))].definition()))

# max s(right: "право (направление)", direction)
sim4 = get_dist_sim(right[1], direction)[1]
print('max s(right: "право (направление)", direction): {}'.format(max(sim4)))
print('closest lemma definition: {}\n'.format(direction[sim4.index(max(sim4))].definition()))

print('max (s(right: "право (человека)", freedom), s(right: "право (человека)", direction)): {}'.format(max(max(sim1), max(sim2))))
print('max (s(right: "право (направление)", freedom), s(right: "право (направление)", direction)): {}'.format(max(max(sim3), max(sim4))))

max s(right: "право (человека)", freedom): 0.09090909090909091
closest lemma definition: the condition of being free; the power to act or speak or think without externally imposed restraints

max s(right: "право (человека)", direction): 0.1111111111111111
closest lemma definition: a general course along which something has a tendency to develop

max s(right: "право (направление)", freedom): 0.07692307692307693
closest lemma definition: the condition of being free; the power to act or speak or think without externally imposed restraints

max s(right: "право (направление)", direction): 0.14285714285714285
closest lemma definition: a line leading to a place or point

max (s(right: "право (человека)", freedom), s(right: "право (человека)", direction)): 0.1111111111111111
max (s(right: "право (направление)", freedom), s(right: "право (направление)", direction)): 0.14285714285714285


Результат тот же (что ожидаемо, потому что используется фактически одна и та же мера). И первое, и второе ближе к direction.

<b>6) Вычислить двумя разными способами расстояние:
d(right: "право (человека)", privilege) и d(organism, whole)
Есть ли разница в расстояниях? Какое из расстояний, по Вашему мнению, в лучшей степени отражает интуитивное представление о семантчиеской близости слов?</b>

А вот здесь же точно речь идет о сходстве/близости значений, а не о стандартном расстоянии? (Потому что стандартное расстояние - кратчайший путь - трудно считать несколькими способами.)


In [50]:
absolute = wordnet.synsets("absolute")
for ss in absolute:
    print(ss, ss.definition())

Synset('absolute.n.01') something that is conceived or that exists independently and not in relation to other things; something that does not depend on anything else and is beyond human control; something that is not relative
Synset('absolute.a.01') perfect or complete or pure
Synset('absolute.s.02') complete and without restriction or qualification; sometimes used informally as intensifiers
Synset('absolute.s.03') not limited by law
Synset('absolute.s.04') expressing finality with no implication of possible change
Synset('absolute.s.05') not capable of being violated or infringed


Посчитаем близость с помощью критериев Path Similarity, Leacock-Chodorow Similarity, Wu-Palmer Similarity, Resnik Similarity, Jiang-Conrath Similarity и Lin Similarity. Для последних трех будем использовать Information Content корпуса Brown.

In [51]:
from nltk.corpus import wordnet_ic
ic = wordnet_ic.ic('ic-brown.dat')

print(right[0].path_similarity(absolute[0]))

#Leacock-Chodorow Similarity
print(right[0].lch_similarity(absolute[0]))

#Wu-Palmer Similarity
print(right[0].wup_similarity(absolute[0]))
    
#Resnik Similarity
print(right[0].res_similarity(absolute[0], ic))
    
#Jiang-Conrath Similarity
print(right[0].jcn_similarity(absolute[0], ic))
    
#Lin Similarity
print(right[0].lin_similarity(absolute[0], ic))


0.3333333333333333
2.538973871058276
0.8888888888888888
6.257652972013772
0.1986316354068356
0.7131329749859666


In [35]:
organism = wordnet.synsets('organism')
for ss in organism:
    print(ss, ss.definition())
    print(ss, ss.examples())

Synset('organism.n.01') a living thing that has (or can develop) the ability to act or function independently
Synset('organism.n.01') []
Synset('organism.n.02') a system considered analogous in structure or function to a living body
Synset('organism.n.02') ['the social organism']


In [36]:
whole = wordnet.synsets('whole', 'n')
for ss in whole:
    print(ss, ss.definition())
    print(ss, ss.examples())

Synset('whole.n.01') all of something including all its component elements or parts
Synset('whole.n.01') ['Europe considered as a whole', 'the whole of American literature']
Synset('whole.n.02') an assemblage of parts that is regarded as a single entity
Synset('whole.n.02') ['how big is that part compared to the whole?', 'the team is a unit']


In [37]:
for ss1 in organism:
    print(ss1.definition())
    for ss2 in whole:
        print(ss2.definition())
        print(ss1.path_similarity(ss2))

        #Leacock-Chodorow Similarity
        print(ss1.lch_similarity(ss2))

        #Wu-Palmer Similarity
        print(ss1.wup_similarity(ss2))

        #Resnik Similarity
        print(ss1.res_similarity(ss2, ic))

        #Jiang-Conrath Similarity
        print(ss1.jcn_similarity(ss2, ic))

        #Lin Similarity
        print(ss1.lin_similarity(ss2, ic))

a living thing that has (or can develop) the ability to act or function independently
all of something including all its component elements or parts
0.07692307692307693
1.072636802264849
0.14285714285714285
-0.0
0.11927453545173074
-0.0
an assemblage of parts that is regarded as a single entity
0.3333333333333333
2.538973871058276
0.8
1.5318337432196856
1.4444255924181877
0.8156763478002892
a system considered analogous in structure or function to a living body
all of something including all its component elements or parts
0.1
1.3350010667323402
0.3076923076923077
0.5962292078977726
0.05928363701403374
0.0660256991738309
an assemblage of parts that is regarded as a single entity
0.125
1.55814461804655
0.2222222222222222
-0.0
0.07444639262208605
-0.0


<style type="text/css">
	table.tableizer-table {
		font-size: 12px;
		border: 1px solid #CCC; 
		font-family: Arial, Helvetica, sans-serif;
	} 
	.tableizer-table td {
		padding: 4px;
		margin: 3px;
		border: 1px solid #CCC;
	}
	.tableizer-table th {
		background-color: #104E8B; 
		color: #FFF;
		font-weight: bold;
	}
</style>
<table class="tableizer-table">
<thead><tr class="tableizer-firstrow"><th>Similarity Index</th><th>right(1)/absolute</th><th>a living thing / all of something</th><th>a living thing / an assemblage</th><th>a system / all of something</th><th>a system / an assemblage</th></tr></thead><tbody>
 <tr><td>Path Similarity</td><td>0.3333333333333333</td><td>0.07692307692</td><td>0.3333333333</td><td>0.1</td><td>0.125</td></tr>
 <tr><td>Leacock-Chodorow Similarity</td><td>2.538973871058276</td><td>1.072636802</td><td>2.538973871</td><td>1.335001067</td><td>1.558144618</td></tr>
 <tr><td>Wu-Palmer Similarity</td><td>0.8888888888888888</td><td>0.1428571429</td><td>0.8</td><td>0.3076923077</td><td>0.2222222222</td></tr>
 <tr><td>Resnik Similarity</td><td>6.257652972013772</td><td>0</td><td>1.531833743</td><td>0.5962292079</td><td>0</td></tr>
 <tr><td>Jiang-Conrath Similarity</td><td>0.1986316354068356</td><td>0.1192745355</td><td>1.444425592</td><td>0.05928363701</td><td>0.07444639262</td></tr>
 <tr><td>Lin Similarity</td><td>0.7131329749859666</td><td>0</td><td>0.8156763478</td><td>0.06602569917</td><td>0</td></tr>
</tbody></table>

Absolute является когипонимом right, так что индексы сходства выше.

Интуитивно кажется, что оба значения организма (a living thing that has (or can develop) the ability to act or function independently; a system considered analogous in structure or function to a living body) должны быть больше похожи на второе значение целого (an assemblage of parts that is regarded as a single entity), чем на первое (all of something including all its component elements or parts). То есть четвертый и шестой столбцы в таблице должны быть больше, чем третий и пятый. Хотя второе значение организма и первое значение целого тоже близки.

В четвертом столбце действительно самые большие индексы среди организма/целого. Только по последним двум индексам эта пара превосходит и right/absolute. 

Показатели в шестом же столбце больше третьего и пятого только для первых двух индексов (по Wu-Palmer Similarity разница небольшая). Первые два индекса, кстати, дают более или менее пропорциональные величины, и сравнение по ним дает одинаковые результаты (это значит, что глубина наших значений в данном случае не влияет на результат).

Если учитывать всё сразу, индексы не очень отражают интуитивное представление о близости слов, а видимо довольно высоки из-за общего гиперонима abstract.